### ISSUE TITLE
how to make the check point while training the model 

### ISSUE THREAD

**Issue by sahasrara62** [explosion/spaCy#1965]
hey i have lot of training data , i want to make some check point  in the code so that 
if in case of my pc failure  , i will be able to resume the training from that point again (the last check point stored while system(pc) got failed .)

## Info about spaCy

* **Models:** en
* **spaCy version:** 2.0.5
* **Python version:** 3.5.4
* **Platform:** Linux-4.13.0-32-generic-x86_64-with-debian-stretch-sid

**Comment by fucking-signup** [comment]
You need to save progress after each iteration like such:

```python
nlp.to_disk(path_to_file)
```

Example:

```python
TRAIN_DATA = [
     ("Uber blew through $1 million a week", {'entities': [(0, 4, 'ORG')]}),
     ("Google rebrands its business apps", {'entities': [(0, 6, "ORG")]})]

nlp = spacy.blank('en')
optimizer = nlp.begin_training()
output_path = '/model'
for i in range(20):
    random.shuffle(TRAIN_DATA)
    for text, annotations in TRAIN_DATA:
        nlp.update([text], [annotations], sgd=optimizer)

    # save to disk after each iteration
    epoch_path = "{0}_epoch_{1}".format(output_path, i)
    nlp.to_disk(epoch_path)

# save the model after the training
nlp.to_disk(output_path)
```

Take a look at `cli/train.py` to see how it's done properly.

**Comment by honnibal** [comment]
You might want to pickle the optimizer too -- that way you'll retain the momentum information.

One thing to note when comparing to the code in `cli/train.py` is that at the end of training, the model is saved within a `with nlp.use_params(optimizer.averages)`. This replaces the final parameters with an exponential moving average, that's tracked during training. This averaging trick (sometimes called Polyak averaging) improves accuracy. However, to train a checkpoint, you don't want to do this --- you want to keep training with the last state of the weights, not the EMA.

**Comment by sahasrara62** [comment]
thanks guys.
